src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,520 @@
+                                                                  ref_column->referenced_column_id(), new_pos_list);
+          output_columns.push_back(new_ref_column);
+        } else {
+          // If there are no Chunks in the input_table, we can't deduce the Table that input_table is referencING to
+          // pos_list will contain only NULL_ROW_IDs anyway, so it doesn't matter which Table the ReferenceColumn that
+          // we output is referencing. HACK, but works fine: we create a dummy table and let the ReferenceColumn ref
+          // it.
+          const auto dummy_table = Table::create_dummy_table(input_table->column_definitions());
+          output_columns.push_back(std::make_shared<ReferenceColumn>(dummy_table, column_id, pos_list));
+        }
+      } else {
+        auto new_ref_column = std::make_shared<ReferenceColumn>(input_table, column_id, pos_list);
+        output_columns.push_back(new_ref_column);
+      }
+    }
+  }
+
+  /**
+  * Turns a pos list that is pointing to reference column entries into a pos list pointing to the original table.
+  * This is done because there should not be any reference columns referencing reference columns.
+  **/
+  std::shared_ptr<PosList> _dereference_pos_list(std::shared_ptr<const Table> input_table, ColumnID column_id,
+                                                 std::shared_ptr<const PosList> pos_list) {
+    // Get all the input pos lists so that we only have to pointer cast the columns once
+    auto input_pos_lists = std::vector<std::shared_ptr<const PosList>>();
+    for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
+      auto b_column = input_table->get_chunk(chunk_id)->get_column(column_id);
+      auto r_column = std::dynamic_pointer_cast<const ReferenceColumn>(b_column);
+      input_pos_lists.push_back(r_column->pos_list());
+    }
+
+    // Get the row ids that are referenced
+    auto new_pos_list = std::make_shared<PosList>();
+    for (const auto& row : *pos_list) {
+      if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
+        new_pos_list->push_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+      } else {
+        new_pos_list->push_back((*input_pos_lists[row.chunk_id])[row.chunk_offset]);
+      }
+    }
+
+    return new_pos_list;
+  }
+
+ public:
+  /**
+  * Executes the MPSMJoin operator.
+  **/
+  std::shared_ptr<const Table> _on_execute() {
+    bool include_null_left = (_mode == JoinMode::Left || _mode == JoinMode::Outer);
+    bool include_null_right = (_mode == JoinMode::Right || _mode == JoinMode::Outer);
+    auto radix_clusterer = RadixClusterSortNUMA<T>(
+        _mpsm_join.input_table_left(), _mpsm_join.input_table_right(), _mpsm_join._column_ids,
+        include_null_left, include_null_right, _cluster_count);
+    // Sort and cluster the input tables
+    auto sort_output = radix_clusterer.execute();
+    _sorted_left_table = std::move(sort_output.clusters_left);
```

Comment:
No need to wrap these four into unique_ptrs, as far as I can tell

Reply:
While we are not exactly on a critical path we prefer not to copy data structures that own data (_sorted_left_table) or of potentially unknown size (null_rows). We could wrap these in shared pointer but prefer unique pointerswhen possible. A reference could do but would cause some restructuring of the code as we could not have these variables as members anymore. The restructuring should then also be done in sort merge join to keep the two classes close. Is this something that should be done?